File: samples/go/agents/signing_and_verifying/test_client.go
package main

import (
	"context"
	"crypto/x509"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"log"
	"net/http"

	"github.com/a2aproject/a2a-go/a2a"
	"github.com/a2aproject/a2a-go/a2aclient"
	"github.com/a2aproject/a2a-go/a2aclient/agentcard"
	"github.com/a2aproject/a2a-go/a2asrv"
)

// keyProvider fetches and parses public key from JKU URL given key ID (kid) and JKU URL.
func keyProvider(kid, jku string) (any, error) {
	if kid == "" {
		return nil, fmt.Errorf("expected kid: string, but got empty string")
	}
	if jku == "" {
		return nil, fmt.Errorf("expected jku: string, but got empty string")
	}

	//nolint:gosec // JKU URL is dynamic by design per RFC 7515
	resp, err := http.Get(jku)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): %w", jku, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): status code %d", jku, resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): failed to read body: %w", jku, err)
	}

	var keys map[string]string
	if unmarshalErr := json.Unmarshal(body, &keys); unmarshalErr != nil {
		return nil, fmt.Errorf("failed to fetch public key from JKU URL (%s): invalid JSON response: %w", jku, unmarshalErr)
	}

	pemStr, ok := keys[kid]
	if !ok || pemStr == "" {
		return nil, fmt.Errorf("invalid JWK Key ID.")
	}

	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM block for key ID %q", kid)
	}

	pubKey, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse PKIX public key for key ID %q: %w", kid, err)
	}

	return pubKey, nil
}

func displayAgentCard(card *a2a.AgentCard) {
	cardJSON, err := json.MarshalIndent(card, "", "  ")
	if err != nil {
		log.Fatalf("Failed to marshal agent card: %v", err)
	}
	log.Println(string(cardJSON))
}

func runTestClient() {
	ctx := context.Background()
	verifyCardSignature := createSignatureVerifier(keyProvider, []string{es256Alg})

	baseURL := serverURL

	log.Printf("Attempting to fetch public agent card from: %s%s", baseURL, a2asrv.WellKnownAgentCardPath)

	// Initialize A2ACardResolver using SDK helper
	resolver := agentcard.NewResolver(http.DefaultClient)
	publicCard, err := resolver.Resolve(ctx, baseURL)
	if err != nil {
		log.Fatalf("Critical error fetching public agent card: %v", err)
	}

	// Pass verifyCardSignature to validate the signature on the public Agent Card
	if verifyErr := verifyCardSignature(publicCard); verifyErr != nil {
		log.Fatalf("Failed to verify public agent card signature: %v", verifyErr)
	}

	log.Println("Successfully fetched public agent card:")

	// Create A2A Client directly via unified card
	client, err := a2aclient.NewFromCard(ctx, publicCard)
	if err != nil {
		log.Fatalf("Failed to create A2A client: %v", err)
	}

	// Pass verifyCardSignature to validate the signature on the extended Agent Card
	extendedCardWithSignature, err := client.GetExtendedAgentCard(ctx, &a2a.GetExtendedAgentCardRequest{})
	if err != nil {
		log.Fatalf("Failed to get extended agent card: %v", err)
	}
	if verifyErr := verifyCardSignature(extendedCardWithSignature); verifyErr != nil {
		log.Fatalf("Failed to verify extended agent card signature: %v", verifyErr)
	}

	log.Println("Successfully fetched extended agent card with signature:")
	displayAgentCard(extendedCardWithSignature)
	log.Println("Signature:")
	sigJSON, err := json.MarshalIndent(extendedCardWithSignature.Signatures, "", "  ")
	if err != nil {
		log.Fatalf("Failed to marshal signature: %v", err)
	}
	log.Println(string(sigJSON))

	// Fetch extended card without signature verification (signature verification is optional on client side)
	extendedCardWithoutSignature, err := client.GetExtendedAgentCard(ctx, &a2a.GetExtendedAgentCardRequest{})
	if err != nil {
		log.Fatalf("Failed to get extended agent card without signature check: %v", err)
	}

	log.Println("Successfully fetched extended agent card without signature:")
	displayAgentCard(extendedCardWithoutSignature)
}

File: samples/js/agents/sign-and-verify-agent-card/test_client.ts
/* eslint-disable n/no-missing-import */
/* eslint-disable n/no-unsupported-features/node-builtins */
/* eslint-disable n/no-process-exit */
import * as crypto from "node:crypto";

import {
    AGENT_CARD_PATH,
    AgentCard,
    verifyAgentCardSignature,
} from "@a2a-js/sdk";
import { ClientFactory, DefaultAgentCardResolver } from "@a2a-js/sdk/client";

async function keyProvider(
    keyId: string,
    jku?: string
): Promise<crypto.KeyObject> {
    /** Fetch and parse public key from JKU URL given key ID (keyId) and JKU URL. */
    if (typeof keyId !== "string" || !keyId) {
        throw new TypeError(
            `Expected keyId: string, but got: ${typeof keyId} (${keyId})`
        );
    }
    if (typeof jku !== "string" || !jku) {
        throw new TypeError(
            `Expected jku: string, but got: ${typeof jku} (${jku})`
        );
    }

    let response: Response;
    try {
        response = await fetch(jku);
        if (!response.ok) {
            throw new Error(`HTTP ${response.status} ${response.statusText}`);
        }
    } catch (err) {
        throw new Error(
            `Failed to fetch public key from JKU URL (${jku}): ${err}`
        );
    }

    const keys = (await response.json()) as Record<string, string>;
    const pemDataStr = keys[keyId];

    if (!pemDataStr) {
        throw new Error("Invalid JWK Key ID.");
    }

    return crypto.createPublicKey(pemDataStr);
}

// Create a verifier function to validate AgentCard JWS signatures
const verifyCardSignature = verifyAgentCardSignature(keyProvider);

async function main(): Promise<void> {
    const baseUrl = "http://localhost:9999";

    // Initialize AgentCardResolver
    const resolver = new DefaultAgentCardResolver();

    console.log(
        `Attempting to fetch public agent card from: ${baseUrl}${AGENT_CARD_PATH}`
    );

    // 1. Fetch public agent card without verifying signature
    try {
        await resolver.resolve(baseUrl);
        console.log(
            "Successfully fetched public agent card without verification."
        );
    } catch (e) {
        console.error("Critical error fetching public agent card:", e);
        throw e;
    }

    // 2. Fetch public agent card and verify signature
    let publicCardVerified: AgentCard;
    try {
        publicCardVerified = await resolver.resolve(baseUrl);
        await verifyCardSignature(publicCardVerified);
        console.log(
            "Successfully fetched public agent card with verification:"
        );
        console.log(JSON.stringify(publicCardVerified, null, 2));
    } catch (e) {
        console.error("Critical error fetching public agent card:", e);
        throw e;
    }

    // Create Client directly via ClientFactory
    const clientFactory = new ClientFactory();
    const client = await clientFactory.createFromAgentCard(publicCardVerified);

    // 3. Fetch extended agent card without signature verification
    const extendedCardUnverified = await client.getAgentCard();
    console.log(
        "Successfully fetched extended agent card without verification:"
    );
    console.log(JSON.stringify(extendedCardUnverified, null, 2));

    // 4. Fetch extended agent card and verify signature
    const extendedCardVerified = await client.getAgentCard(
        undefined,
        verifyCardSignature
    );
    console.log("Successfully fetched extended agent card with verification:");
    console.log(JSON.stringify(extendedCardVerified, null, 2));
}

main().catch((err) => {
    console.error("Error in main:", err);
    process.exit(1);
});
